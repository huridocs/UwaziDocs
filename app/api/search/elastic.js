import elasticSearch from 'elasticsearch';

const elastic = new elasticSearch.Client({
  host: process.env.ELASTICSEARCH_URL || 'http://localhost:9200',
  maxRetries: 5,
  requestTimeout: 60000,
});

export default elastic;
