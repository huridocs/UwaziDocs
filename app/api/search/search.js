import { comonProperties, defaultFilters, allUniqueProperties, textFields } from 'shared/comonProperties';
import { detect as detectLanguage } from 'shared/languages';
import { index as elasticIndex } from 'api/config/elasticIndexes';
import languages from 'shared/languagesList';
import dictionariesModel from 'api/thesauris/dictionariesModel';

import documentQueryBuilder from './documentQueryBuilder';
import elastic from './elastic';
import entities from '../entities';
import templatesModel from '../templates';

function processFiltes(filters, properties) {
  return Object.keys(filters || {}).map((propertyName) => {
    const property = properties.find(p => p.name === propertyName);
    let type = 'text';
    if (property.type === 'date' || property.type === 'multidate' || property.type === 'numeric') {
      type = 'range';
    }
    if (property.type === 'select' || property.type === 'multiselect' || property.type === 'relationship') {
      type = 'multiselect';
    }
    if (property.type === 'nested') {
      type = 'nested';
    }
    if (property.type === 'multidaterange' || property.type === 'daterange') {
      type = 'nestedrange';
    }
    return Object.assign(property, { value: filters[property.name], type });
  });
}

function filtersBasedOnSearchTerm(properties, entitiesMatchedByTitle, dictionariesMatchByLabel) {
  if (!entitiesMatchedByTitle.length && !dictionariesMatchByLabel.length) {
    return [];
  }
  let values = entitiesMatchedByTitle.map(item => item.sharedId);
  values = values.concat(dictionariesMatchByLabel.map(dictionary => dictionary.values.id));
  return properties.map((prop) => {
    if (prop.type === 'select' || prop.type === 'multiselect') {
      return { name: prop.name, value: { values }, type: 'multiselect' };
    }
  }).filter(f => f);
}

function agregationProperties(properties) {
  return properties
  .filter(property => property.type === 'select' ||
    property.type === 'multiselect' ||
    property.type === 'relationship' ||
    property.type === 'nested')
  .map((property) => {
    if (property.type === 'nested') {
      return { name: property.name, nested: true, nestedProperties: property.nestedProperties };
    }
    return { name: property.name, nested: false, content: property.content };
  });
}

function searchGeolocation(documentsQuery, filteringTypes, templates) {
  documentsQuery.limit(9999);
  const geolocationProperties = [];
  templates.forEach((template) => {
    template.properties.forEach((prop) => {
      if (prop.type === 'geolocation') {
        geolocationProperties.push(prop.name);
      }
    });
  });
  documentsQuery.hasMetadataProperties(geolocationProperties);
  const selectProps = geolocationProperties.map(p => `metadata.${p}`)
  .concat(['title', 'template', 'sharedId', 'language']);
  documentsQuery.select(selectProps);
}

const search = {
  search(query, language, user) {
    let searchEntitiesbyTitle = Promise.resolve([]);
    let searchDictionariesByTitle = Promise.resolve([]);
    if (query.searchTerm) {
      searchEntitiesbyTitle = entities.get({ $text: { $search: query.searchTerm }, language });
      const regexp = `.*${query.searchTerm}.*`;
      searchDictionariesByTitle = dictionariesModel.db.aggregate([
        { $match: { 'values.label': { $regex: regexp, $options: 'i' } } },
        { $unwind: '$values' },
        { $match: { 'values.label': { $regex: regexp, $options: 'i' } } }
      ]);
    }

    return Promise.all([templatesModel.get(), searchEntitiesbyTitle, searchDictionariesByTitle, dictionariesModel.get()])
    .then(([templates, entitiesMatchedByTitle, dictionariesMatchByLabel, dictionaries]) => {
      const textFieldsToSearch = query.fields || textFields(templates).map(prop => `metadata.${prop.name}`).concat(['title', 'fullText']);
      const documentsQuery = documentQueryBuilder()
      .fullTextSearch(query.searchTerm, textFieldsToSearch, 2)
      .filterByTemplate(query.types)
      .filterById(query.ids)
      .language(language);

      if (query.sort) {
        documentsQuery.sort(query.sort, query.order);
      }

      if (query.from) {
        documentsQuery.from(query.from);
      }

      if (query.limit) {
        documentsQuery.limit(query.limit);
      }

      if (query.includeUnpublished && user) {
        documentsQuery.includeUnpublished();
      }

      if (query.unpublished && user) {
        documentsQuery.unpublished();
      }

      const allTemplates = templates.map(t => t._id);
      const filteringTypes = query.types && query.types.length ? query.types : allTemplates;
      let properties = comonProperties(templates, filteringTypes);
      properties = !query.types || !query.types.length ? defaultFilters(templates) : properties;

      if (query.allAggregations) {
        properties = allUniqueProperties(templates);
      }

      const aggregations = agregationProperties(properties);
      const filters = processFiltes(query.filters, properties);
      const textSearchFilters = filtersBasedOnSearchTerm(allUniqueProperties(templates), entitiesMatchedByTitle, dictionariesMatchByLabel);

      documentsQuery.filterMetadataByFullText(textSearchFilters);
      documentsQuery.filterMetadata(filters);
      documentsQuery.aggregations(aggregations, dictionaries);

      if (query.geolocation) {
        searchGeolocation(documentsQuery, filteringTypes, templates);
      }

      return elastic.search({ index: elasticIndex, body: documentsQuery.query() })
      .then((response) => {
        const rows = response.hits.hits.map((hit) => {
          const result = hit._source;
          result._explanation = hit._explanation;
          result.snippets = [];
          if (hit.inner_hits && hit.inner_hits.fullText.hits.hits.length) {
            const regex = /\[\[(\d+)\]\]/g;

            const highlights = hit.inner_hits.fullText.hits.hits[0].highlight;

            result.snippets = highlights[Object.keys(highlights)[0]].map((snippet) => {
              const matches = regex.exec(snippet);
              return {
                text: snippet.replace(regex, ''),
                page: matches ? Number(matches[1]) : 0
              };
            });
          }
          result._id = hit._id;
          return result;
        });
        Object.keys(response.aggregations.all).forEach((aggregationKey) => {
          const aggregation = response.aggregations.all[aggregationKey];
          if (aggregation.buckets && !Array.isArray(aggregation.buckets)) {
            aggregation.buckets = Object.keys(aggregation.buckets).map(key => Object.assign({ key }, aggregation.buckets[key]));
          }
          response.aggregations.all[aggregationKey] = aggregation;
        });

        return { rows, totalRows: response.hits.total, aggregations: response.aggregations };
      });
    });
  },

  getUploadsByUser(user, language) {
    return entities.get({ user: user._id, language, published: false });
  },

  searchSnippets(searchTerm, sharedId, language) {
    const query = documentQueryBuilder()
    .fullTextSearch(searchTerm, ['fullText'], 9999)
    .includeUnpublished()
    .filterById(sharedId)
    .language(language)
    .query();

    return elastic.search({ index: elasticIndex, body: query })
    .then((response) => {
      if (response.hits.hits.length === 0) {
        return [];
      }

      if (!response.hits.hits[0].inner_hits) {
        return [];
      }

      const highlights = response.hits.hits[0].inner_hits.fullText.hits.hits[0].highlight;

      const regex = /\[\[(\d+)\]\]/g;
      return highlights[Object.keys(highlights)[0]].map((snippet) => {
        const matches = regex.exec(snippet);
        return {
          text: snippet.replace(regex, ''),
          page: matches ? Number(matches[1]) : 0
        };
      });
    });
  },

  index(_entity) {
    const entity = Object.assign({}, _entity);
    const id = entity._id.toString();
    delete entity._id;
    delete entity._rev;
    delete entity.pdfInfo;

    const body = entity;
    let fullTextIndex = Promise.resolve();
    if (entity.fullText) {
      const fullText = {};
      let language;
      if (!entity.file || entity.file && !entity.file.language) {
        language = detectLanguage(entity.fullText);
      }
      if (entity.file && entity.file.language) {
        language = languages(entity.file.language);
      }

      fullText[`fullText_${language}`] = entity.fullText;
      fullTextIndex = elastic.index({ index: elasticIndex, type: 'fullText', parent: id, body: fullText, id: `${id}_fullText` });
      delete entity.fullText;
    }
    return Promise.all([
      elastic.index({ index: elasticIndex, type: 'entity', id, body }),
      fullTextIndex
    ]);
  },

  bulkIndex(docs, _action = 'index') {
    const type = 'entity';
    const body = [];
    docs.forEach((doc) => {
      let _doc = doc;
      const id = doc._id.toString();
      delete doc._id;
      delete doc._rev;
      delete doc.pdfInfo;

      let action = {};
      action[_action] = { _index: elasticIndex, _type: type, _id: id };
      if (_action === 'update') {
        _doc = { doc: _doc };
      }

      body.push(action);
      body.push(_doc);

      if (doc.fullText) {
        action = {};
        action[_action] = { _index: elasticIndex, _type: 'fullText', parent: id, _id: `${id}_fullText` };
        body.push(action);

        const fullText = {};
        let language;
        if (!doc.file || doc.file && !doc.file.language) {
          language = detectLanguage(doc.fullText);
        }
        if (doc.file && doc.file.language) {
          language = languages(doc.file.language);
        }
        fullText[`fullText_${language}`] = doc.fullText;
        body.push(fullText);
        delete doc.fullText;
      }
    });

    return elastic.bulk({ body })
    .then((res) => {
      if (res.items) {
        res.items.forEach((f) => {
          if (f.index.error) {
            console.log(`ERROR Failed to index document ${f.index._id}: ${JSON.stringify(f.index.error, null, ' ')}`);
          }
        });
      }
      return res;
    });
  },

  delete(entity) {
    const id = entity._id.toString();
    return elastic.delete({ index: elasticIndex, type: 'entity', id });
  }
};

export default search;
