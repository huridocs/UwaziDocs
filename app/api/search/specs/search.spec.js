/* eslint-disable max-nested-callbacks */
import { index as elasticIndex } from 'api/config/elasticIndexes';
import { search, documentQueryBuilder, elastic } from 'api/search';
import { catchErrors } from 'api/utils/jasmineHelpers';
import db from 'api/utils/testing_db';
import instanceElasticTesting from 'api/utils/elastic_testing';
import elasticResult from './elasticResult';
import elasticFixtures, { ids } from './fixtures_elastic';

describe('search', () => {
  let result;
  const elasticTesting = instanceElasticTesting('search_index_test');

  beforeAll((done) => {
    result = elasticResult().withDocs(
    [
      { title: 'doc1',
        _id: 'id1',
        snippets: {
          hits: {
            hits: [
              {
                highlight: {
                  fullText: []
                }
              }
            ]
          }
        }
      }
    ])
    .toObject();

    db.clearAllAndLoad(elasticFixtures)
    .then(() => elasticTesting.reindex())
    .then(done)
    .catch(catchErrors(done));
  });

  afterAll((done) => {
    db.disconnect().then(done);
  });

  describe('getUploadsByUser', () => {
    it('should request all unpublished entities or documents for the user', (done) => {
      const user = { _id: ids.userId };
      search.getUploadsByUser(user, 'en')
      .then((response) => {
        expect(response.length).toBe(1);
        expect(response[0].title).toBe('metadata6');
        done();
      })
      .catch(catchErrors(done));
    });
  });

  describe('search', () => {
    describe('searchSnippets', () => {
      it('perform a search on fullText of the document passed and return the snippets', (done) => {
        search.searchSnippets('spanish', ids.batmanFinishes, 'es')
        .then((snippets) => {
          expect(snippets.length).toBe(1);
          expect(snippets[0].page).toBe(34);
          expect(snippets[0].text).toMatch('spanish');
          expect(snippets[0].text).not.toMatch('[[34]]');
          done();
        })
        .catch(catchErrors(done));
      });

      it('should perform the search on unpublished documents also', (done) => {
        search.searchSnippets('unpublished', 'unpublishedSharedId', 'en')
        .then((snippets) => {
          expect(snippets.length).toBe(1);
          done();
        })
        .catch(catchErrors(done));
      });

      describe('when document is not matched', () => {
        it('should return empty array', (done) => {
          search.searchSnippets('not matching string', ids.batmanFinishes, 'es')
          .then((snippets) => {
            expect(snippets.length).toBe(0);
            done();
          })
          .catch(catchErrors(done));
        });
      });

      describe('when searchTerm is empty', () => {
        it('should return empty array', (done) => {
          search.searchSnippets('', ids.batmanFinishes, 'es')
          .then((snippets) => {
            expect(snippets.length).toBe(0);
            done();
          })
          .catch(catchErrors(done));
        });
      });
    });

    it('should perform a fullTextSearch on passed fields', (done) => {
      Promise.all([
        search.search({ searchTerm: 'spanish', fields: ['title'] }, 'es'),
        search.search({ searchTerm: 'Batman', fields: ['title'] }, 'es')
      ])
      .then(([resultsNotFound, resultsFound]) => {
        expect(resultsNotFound.rows.length).toBe(0);
        expect(resultsFound.rows.length).toBe(2);
        done();
      })
      .catch(catchErrors(done));
    });

    it('should perform a fullTextSearch on fullText and title', (done) => {
      Promise.all([
        search.search({ searchTerm: 'spanish' }, 'es'),
        search.search({ searchTerm: 'english' }, 'es'),
        search.search({ searchTerm: 'english' }, 'en'),
        search.search({ searchTerm: 'Batman finishes' }, 'en'),
        search.search({ searchTerm: 'Batman begins' }, 'es'),
        search.search({ searchTerm: 'Batman' }, 'en')
      ])
      .then(([spanish, none, english, batmanFinishes, batmanBegins, batman]) => {
        expect(english.rows.find(r => r.snippets[0].text.match('<b>english</b> document <b>english</b>')).snippets[0].page).toBe(12);
        expect(english.rows.find(r => r.snippets[0].text.match('<b>english</b> another')).snippets[0].page).toBe(2);
        expect(english.rows.length).toBe(2);

        expect(spanish.rows.length).toBe(1);
        expect(none.rows.length).toBe(0);
        expect(batmanFinishes.rows.length).toBe(1);
        expect(batmanBegins.rows.length).toBe(1);
        expect(batman.rows.length).toBe(2);
        done();
      })
      .catch(catchErrors(done));
    });

    it('should return aggregations when searching by 2 terms', (done) => {
      search.search({ searchTerm: 'english document' }, 'es')
      .then((response) => {
        const aggregation = response.aggregations.all.types.buckets.find(bucket => bucket.key === ids.template1.toString());
        expect(aggregation.filtered.doc_count).toBe(1);
        done();
      })
      .catch(catchErrors(done));
    });

    it('should convert aggregation buckets that are objects to arrays', (done) => {
      result.aggregations.all = {
        dictionaryWithGroups: {
          buckets: {
            a: { doc_count: 2, filtered: { doc_count: 1 } },
            b: { doc_count: 2, filtered: { doc_count: 1 } }
          }
        }
      };
      spyOn(elastic, 'search').and.returnValue(Promise.resolve(result));
      search.search({ searchTerm: '', geolocation: true }, 'en')
      .then((response) => {
        const expectedBuckets = [{ key: 'a', doc_count: 2, filtered: { doc_count: 1 } }, { key: 'b', doc_count: 2, filtered: { doc_count: 1 } }];
        expect(response.aggregations.all.dictionaryWithGroups.buckets).toEqual(expectedBuckets);
        done();
      });
    });

    it('should match entities related somehow with other entities with a title that is the search term', (done) => {
      search.search({ searchTerm: 'egypt' }, 'en')
      .then(({ rows }) => {
        expect(rows.length).toBe(3);
        const country = rows.find(_result => _result.sharedId === 'abc123');
        const entityWithEgypt = rows.find(_result => _result.sharedId === 'entityWithEgypt');
        const entityWithEgyptDictionary = rows.find(_result => _result.sharedId === 'entityWithEgyptDictionary');
        expect(country).toBeDefined();
        expect(entityWithEgypt).toBeDefined();
        expect(entityWithEgyptDictionary).toBeDefined();
        done();
      })
      .catch(catchErrors(done));
    });

    it('should filter by templates', (done) => {
      Promise.all([
        search.search({ types: [ids.template1] }, 'es'),
        search.search({ types: [ids.template2] }, 'es'),
        search.search({ types: [ids.template1] }, 'en'),
        search.search({ types: [ids.template1, ids.template2] }, 'en'),
        search.search({ types: ['missing'] }, 'en'),
        search.search({ types: [ids.template1, 'missing'] }, 'en')
      ])
      .then(([template1es, template2es, template1en, allTemplatesEn, onlyMissing, template1AndMissing]) => {
        expect(template1es.rows.length).toBe(2);
        expect(template1en.rows.length).toBe(2);
        expect(template2es.rows.length).toBe(1);
        expect(allTemplatesEn.rows.length).toBe(3);
        expect(onlyMissing.rows.length).toBe(2);
        expect(template1AndMissing.rows.length).toBe(4);
        done();
      })
      .catch(catchErrors(done));
    });

    it('should allow searching only within specific Ids', (done) => {
      Promise.all([
        search.search({ ids: [ids.batmanBegins] }, 'es'),
        search.search({ ids: ids.batmanBegins }, 'en'),
        search.search({ ids: [ids.batmanFinishes, ids.batmanBegins] }, 'en')
      ])
      .then(([es, en, both]) => {
        expect(es.rows.length).toBe(1);
        expect(es.rows[0].title).toBe('Batman begins es');
        expect(en.rows.length).toBe(1);
        expect(en.rows[0].title).toBe('Batman begins en');
        expect(both.rows.length).toBe(2);
        expect(both.rows.find(r => r.title === 'Batman finishes en')).not.toBe(null);
        expect(both.rows.find(r => r.title === 'Batman begins en')).not.toBe(null);
        done();
      })
      .catch(catchErrors(done));
    });

    it('should filter by metadata, and return template aggregations based on the filter the language and the published status', (done) => {
      Promise.all([
        search.search({ types: [ids.templateMetadata1, ids.templateMetadata2], filters: { field1: 'joker' } }, 'en'),
        search.search({ types: [ids.templateMetadata1, ids.templateMetadata2], unpublished: true }, 'en', { _id: 'user' })
      ])
      .then(([joker, unpublished]) => {
        expect(joker.rows.length).toBe(2);

        const typesAggs = joker.aggregations.all.types.buckets;
        expect(typesAggs.find(a => a.key === ids.templateMetadata1).filtered.doc_count).toBe(2);
        expect(typesAggs.find(a => a.key === ids.templateMetadata2).filtered.doc_count).toBe(0);

        const unpublishedAggs = unpublished.aggregations.all.types.buckets;
        expect(unpublishedAggs.find(a => a.key === ids.templateMetadata1).filtered.doc_count).toBe(1);
        expect(unpublishedAggs.find(a => a.key === ids.templateMetadata2).filtered.doc_count).toBe(0);
        done();
      })
      .catch(catchErrors(done));
    });

    it('should filter by fullText, and return template aggregations based on the filter the language and the published status', (done) => {
      Promise.all([
        search.search({ searchTerm: 'spanish' }, 'es')
      ])
      .then(([matches]) => {
        const matchesAggs = matches.aggregations.all.types.buckets;
        expect(matchesAggs.find(a => a.key === ids.template1).filtered.doc_count).toBe(1);
        expect(matchesAggs.find(a => a.key === ids.template2).filtered.doc_count).toBe(0);
        expect(matchesAggs.find(a => a.key === ids.templateMetadata1).filtered.doc_count).toBe(0);
        expect(matchesAggs.find(a => a.key === ids.templateMetadata2).filtered.doc_count).toBe(0);
        done();
      })
      .catch(catchErrors(done));
    });

    describe('when the query is for geolocation', () => {
      it('should set size to 9999', (done) => {
        spyOn(elastic, 'search').and.returnValue(Promise.resolve(result));
        search.search({ searchTerm: '', geolocation: true }, 'en')
        .then(() => {
          const elasticQuery = elastic.search.calls.argsFor(0)[0].body;
          expect(elasticQuery.size).toBe(9999);
          done();
        });
      });

      it('should only get entities with geolocation fields ', (done) => {
        search.search({ searchTerm: '', geolocation: true }, 'en')
        .then((entities) => {
          expect(entities.rows.length).toBe(3);
          done();
        });
      });

      it('should only select title and geolocation fields ', (done) => {
        search.search({ searchTerm: '', geolocation: true }, 'en')
        .then(({ rows }) => {
          expect(rows[0].title).toBeDefined();
          expect(rows[0].template).toBeDefined();
          expect(rows[0].sharedId).toBeDefined();
          expect(rows[0].language).toBeDefined();
          expect(Object.keys(rows[0].metadata).length).toBe(1);
          done();
        });
      });
    });

    describe('select aggregations', () => {
      it('should return aggregations of select fields when filtering by types', (done) => {
        Promise.all([
          search.search({ types: [ids.templateMetadata1] }, 'en'),
          search.search({ types: [ids.templateMetadata2] }, 'en'),
          search.search({ types: [ids.templateMetadata1, ids.templateMetadata2] }, 'en'),
          search.search({ types: [ids.templateMetadata1], unpublished: true }, 'en', { _id: 'user' })
        ])
        .then(([template1, template2, both, template1Unpublished]) => {
          const template1Aggs = template1.aggregations.all.select1.buckets;
          expect(template1Aggs.find(a => a.key === 'selectValue1').filtered.doc_count).toBe(2);
          expect(template1Aggs.find(a => a.key === 'selectValue2').filtered.doc_count).toBe(1);
          expect(template1Aggs.find(a => a.key === 'missing').filtered.doc_count).toBe(0);

          const template2Aggs = template2.aggregations.all.select1.buckets;
          expect(template2Aggs.find(a => a.key === 'selectValue1').filtered.doc_count).toBe(0);
          expect(template2Aggs.find(a => a.key === 'selectValue2').filtered.doc_count).toBe(1);
          expect(template2Aggs.find(a => a.key === 'missing').filtered.doc_count).toBe(1);

          const bothAggs = both.aggregations.all.select1.buckets;
          expect(bothAggs.find(a => a.key === 'selectValue1').filtered.doc_count).toBe(2);
          expect(bothAggs.find(a => a.key === 'selectValue2').filtered.doc_count).toBe(2);
          expect(bothAggs.find(a => a.key === 'missing').filtered.doc_count).toBe(1);

          const template1UnpubishedAggs = template1Unpublished.aggregations.all.select1.buckets;
          expect(template1UnpubishedAggs.find(a => a.key === 'selectValue1').filtered.doc_count).toBe(0);
          expect(template1UnpubishedAggs.find(a => a.key === 'selectValue2').filtered.doc_count).toBe(0);
          done();
        })
        .catch(catchErrors(done));
      });
    });

    describe('multiselect aggregations', () => {
      it('should return aggregations of multiselect fields', (done) => {
        Promise.all([
          search.search({ types: [ids.templateMetadata1] }, 'en'),
          search.search({ types: [ids.templateMetadata2] }, 'en'),
          search.search({ types: [ids.templateMetadata1, ids.templateMetadata2] }, 'en'),
          search.search({ filters: { multiselect1: { values: ['multiValue2'], and: false } },
          types: [ids.templateMetadata1, ids.templateMetadata2] }, 'en')
        ])
        .then(([template1, template2, both, filtered]) => {
          const template1Aggs = template1.aggregations.all.multiselect1.buckets;
          expect(template1Aggs.find(a => a.key === 'multiValue1').filtered.doc_count).toBe(2);
          expect(template1Aggs.find(a => a.key === 'multiValue2').filtered.doc_count).toBe(2);

          const template2Aggs = template2.aggregations.all.multiselect1.buckets;
          expect(template2Aggs.find(a => a.key === 'multiValue1').filtered.doc_count).toBe(0);
          expect(template2Aggs.find(a => a.key === 'multiValue2').filtered.doc_count).toBe(1);

          const bothAggs = both.aggregations.all.multiselect1.buckets;
          expect(bothAggs.find(a => a.key === 'multiValue1').filtered.doc_count).toBe(2);
          expect(bothAggs.find(a => a.key === 'multiValue2').filtered.doc_count).toBe(3);

          const filteredAggs = filtered.aggregations.all.multiselect1.buckets;
          const templateAggs = filtered.aggregations.all.types.buckets;
          expect(filteredAggs.find(a => a.key === 'multiValue1').filtered.doc_count).toBe(2);
          expect(filteredAggs.find(a => a.key === 'multiValue2').filtered.doc_count).toBe(3);
          expect(templateAggs.find(a => a.key === ids.template1).filtered.doc_count).toBe(0);
          expect(templateAggs.find(a => a.key === ids.template2).filtered.doc_count).toBe(0);

          done();
        })
        .catch(catchErrors(done));
      });

      describe('allAggregations', () => {
        it('should return all aggregations', async () => {
          const allAggregations = await search.search({ allAggregations: true }, 'en');
          const aggregationsIncluded = (Object.keys(allAggregations.aggregations.all));
          expect(aggregationsIncluded).toMatchSnapshot();
        });
      });

      describe('AND flag', () => {
        it('should restrict the results to those who have all values of the filter', (done) => {
          search.search(
            {
              filters: { multiselect1: { values: ['multiValue1', 'multiValue2'], and: true } },
              types: [ids.templateMetadata1, ids.templateMetadata2]
            }, 'en')
          .then((filtered) => {
            const filteredAggs = filtered.aggregations.all.multiselect1.buckets;
            const templateAggs = filtered.aggregations.all.types.buckets;
            expect(filteredAggs.find(a => a.key === 'multiValue1').filtered.doc_count).toBe(1);
            expect(filteredAggs.find(a => a.key === 'multiValue2').filtered.doc_count).toBe(1);
            expect(templateAggs.find(a => a.key === ids.templateMetadata1).filtered.doc_count).toBe(1);

            done();
          })
          .catch(catchErrors(done));
        });
      });

      describe('nested', () => {
        it('should search by nested and calculate nested aggregations of fields when filtering by types', (done) => {
          Promise.all([
            search.search({ types: [ids.templateMetadata2] }, 'en'),
            search.search({ types: [ids.templateMetadata1, ids.templateMetadata2],
              filters: { nestedField: { properties: { nested1: { any: true } } } } }, 'en')
          ])
          .then(([template2NestedAggs, nestedSearchFirstLevel]) => {
            const nestedAggs = template2NestedAggs.aggregations.all.nestedField.nested1.buckets;
            expect(template2NestedAggs.rows.length).toBe(2);
            expect(nestedAggs.find(a => a.key === '3').filtered.total.filtered.doc_count).toBe(1);
            expect(nestedAggs.find(a => a.key === '4').filtered.total.filtered.doc_count).toBe(1);
            expect(nestedAggs.find(a => a.key === '6').filtered.total.filtered.doc_count).toBe(1);
            expect(nestedAggs.find(a => a.key === '7').filtered.total.filtered.doc_count).toBe(1);
            expect(nestedAggs.find(a => a.key === '5').filtered.total.filtered.doc_count).toBe(2);

            const bothTemplatesAggs = nestedSearchFirstLevel.aggregations.all.nestedField.nested1.buckets;
            expect(nestedSearchFirstLevel.rows.length).toBe(3);
            expect(bothTemplatesAggs.find(a => a.key === '1').filtered.total.filtered.doc_count).toBe(1);
            expect(bothTemplatesAggs.find(a => a.key === '2').filtered.total.filtered.doc_count).toBe(1);
            expect(bothTemplatesAggs.find(a => a.key === '3').filtered.total.filtered.doc_count).toBe(2);
            expect(bothTemplatesAggs.find(a => a.key === '4').filtered.total.filtered.doc_count).toBe(1);
            expect(bothTemplatesAggs.find(a => a.key === '6').filtered.total.filtered.doc_count).toBe(1);
            expect(bothTemplatesAggs.find(a => a.key === '7').filtered.total.filtered.doc_count).toBe(1);
            expect(bothTemplatesAggs.find(a => a.key === '5').filtered.total.filtered.doc_count).toBe(2);
            done();
          })
          .catch(catchErrors(done));
        });

        it('should search second level of nested field', (done) => {
          Promise.all([
            search.search({ types: [ids.templateMetadata1, ids.templateMetadata2],
              filters: {
                nestedField: { properties: { nested1: { values: ['1'] } } }
              } }, 'en'),
            search.search({ types: [ids.templateMetadata1, ids.templateMetadata2],
              filters: {
                nestedField: { properties: { nested1: { values: ['2'] } } }
              } }, 'en'),
            search.search({ types: [ids.templateMetadata1, ids.templateMetadata2],
              filters: {
                nestedField: { properties: { nested1: { values: ['3'] } } }
              } }, 'en'),
            search.search({ types: [ids.templateMetadata1, ids.templateMetadata2],
              filters: {
                nestedField: { properties: { nested1: { values: ['3', '5'] } } }
              } }, 'en')
          ])
          .then(([value1, value2, value3, value35]) => {
            expect(value1.rows.length).toBe(1);
            expect(value1.rows[0].title).toBe('metadata1');

            expect(value2.rows.length).toBe(1);
            expect(value2.rows[0].title).toBe('metadata1');

            expect(value3.rows.length).toBe(2);
            expect(value3.rows.find(r => r.title === 'metadata1')).toBeDefined();
            expect(value3.rows.find(r => r.title === ' Metadáta4')).toBeDefined();

            expect(value35.rows.length).toBe(1);
            expect(value35.rows.find(r => r.title === ' Metadáta4')).toBeDefined();

            done();
          })
          .catch(catchErrors(done));
        });

        describe('strict nested filter', () => {
          it('should return only results with values selected in the same key', (done) => {
            Promise.all([
              search.search({ types: [ids.templateMetadata1, ids.templateMetadata2],
                filters: {
                  nestedField: { properties: { nested1: { values: ['1', '5'] } }, strict: true }
                } }, 'en'),
              search.search({ types: [ids.templateMetadata1, ids.templateMetadata2],
                filters: {
                  nestedField: { properties: { nested1: { values: ['1', '2'] } }, strict: true }
                } }, 'en')
            ])
            .then(([strict15, strict12]) => {
              expect(strict15.rows.length).toBe(0);
              expect(strict12.rows.length).toBe(1);
              done();
            })
            .catch(catchErrors(done));
          });
        });
      });
    });

    it('should sort (ignoring case and leading whitespaces) if sort is present', (done) => {
      Promise.all([
        search.search({ types: [ids.templateMetadata1, ids.templateMetadata2], order: 'asc', sort: 'title' }, 'en'),
        search.search({ types: [ids.templateMetadata1, ids.templateMetadata2], order: 'desc', sort: 'title' }, 'en')
      ])
      .then(([asc, desc]) => {
        expect(asc.rows[0].title).toBe('metadata1');
        expect(asc.rows[1].title).toBe('Metadata2');
        expect(asc.rows[2].title).toBe('metádata3');
        expect(asc.rows[3].title).toBe(' Metadáta4');

        expect(desc.rows[0].title).toBe('metadata5');
        expect(desc.rows[1].title).toBe(' Metadáta4');
        expect(desc.rows[2].title).toBe('metádata3');
        expect(desc.rows[3].title).toBe('Metadata2');
        done();
      })
      .catch(catchErrors(done));
    });

    it('should allow including unpublished documents if user', (done) => {
      spyOn(elastic, 'search').and.returnValue(new Promise(resolve => resolve(result)));
      search.search({
        searchTerm: 'searchTerm',
        includeUnpublished: true
      }, 'es', 'user')
      .then(() => {
        const expectedQuery = documentQueryBuilder()
        .fullTextSearch('searchTerm', ['metadata.field1', 'metadata.field2', 'metadata.field3', 'title', 'fullText'], 2)
        .includeUnpublished()
        .language('es')
        .query();

        expect(elastic.search).toHaveBeenCalledWith({ index: elasticIndex, body: expectedQuery });
        done();
      });
    });

    it('should not include unpublished documents if no user', (done) => {
      spyOn(elastic, 'search').and.returnValue(new Promise(resolve => resolve(result)));
      search.search({
        searchTerm: 'searchTerm',
        includeUnpublished: true
      }, 'es')
      .then(() => {
        const expectedQuery = documentQueryBuilder()
        .fullTextSearch('searchTerm', ['metadata.field1', 'metadata.field2', 'metadata.field3', 'title', 'fullText'], 2)
        .language('es')
        .query();

        expect(elastic.search).toHaveBeenCalledWith({ index: elasticIndex, body: expectedQuery });
        done();
      });
    });
  });
});
