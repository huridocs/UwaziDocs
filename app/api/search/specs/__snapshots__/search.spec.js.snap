// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search search multiselect aggregations allAggregations should return all aggregations 1`] = `
Array [
  "doc_count",
  "select1",
  "types",
  "multiselect1",
  "nestedField",
]
`;
